### PR TITLE
59 allow training hours to be submitted

### PR DESCRIPTION
**Pull request by changukshin**
Not ready to merge.
Need below jobs to do.

- [x] live test for clockify
- [x] live test for timesheet

I will tag and request review for you when I pass all of above tests.

### CODE REVIEW

**Comment by changukshin** [comment]
All settled. Review required.
@ThorpeJosh @siewyeng @lamchuntat

File: tp_timesheet/tests/test_clockify.py [hunk 1]
```diff
@@ -79,15 +77,18 @@ def assert_number_of_entries(test_date, number_of_entries):
     assert_number_of_entries(test_date, 0)
 
     # Add a clockify entry with 1 hour
```
Comment (siewyeng) [inline tp_timesheet/tests/test_clockify.py:79]: small comment, could you change this commend of 1 hour too

Comment (changukshin) [reply tp_timesheet/tests/test_clockify.py]: Resolved with 8be4f69. Thanks for your review. @siewyeng

**Review by siewyeng — APPROVED** [review]

File: .pylintrc [hunk 2]
```diff
@@ -336,7 +336,7 @@ indent-after-paren=4
 indent-string='    '
 
 # Maximum number of characters on a single line.
-max-line-length=120
+max-line-length=130
```
Comment (ThorpeJosh) [inline .pylintrc:339]: I'd try to avoid going over 120.
The python line length standards I've seen are 88, 100 or 120. 120 is the default width of most modern shell windows when you open them.

File: tp_timesheet/__main__.py [hunk 3]
```diff
@@ -76,13 +70,25 @@ def parse_args():
...
-        type=str,
-        default="live",
-        choices=["holiday", "live", "OOO", "training"],
+        action="append",
+        nargs=2,
+        metavar=("task", "hour"),
         help="The type of task for the clockify submission. Specify task name "
-        + "if it is anything other than 'live'. Put the task name (e.g., training, OOO, holiday)",
+        + "if it is anything other than 'live'. Put the task name(idle, training, issue, OOO, holiday).\n"
+        + 'Passing multiple task-hour pair is also acceptable. (ex, "--task live 4 --task OOO 4" for afternoon OOO.)',
```
Comment (ThorpeJosh) [inline tp_timesheet/__main__.py:78]:
```suggestion
        + 'Passing multiple task-hour pair is also acceptable. (E.g. "--task live 4 --task OOO 4" for afternoon OOO.)',
```
`e.g.` means "for example"

File: tp_timesheet/__main__.py [hunk 4]
```diff
@@ -76,13 +70,25 @@ def parse_args():
...
     )
-    return parser.parse_args()
+    args = parser.parse_args()
+
+    # postprocessing args
+    args.task = [["live", "8"]] if not args.task else args.task  # default value
+    args.task = {k: int(v) for (k, v) in args.task}
```
Comment (ThorpeJosh) [inline tp_timesheet/__main__.py:84]:
```suggestion
    args.task = {task_name: int(hours) for (task_name, hours) in args.task}
```
Just to be more obvious about what the tuple contents are.

File: tp_timesheet/__main__.py [hunk 5]
```diff
@@ -76,13 +70,25 @@ def parse_args():
...
+    if sum(args.task.values()) != 8:
+        raise ValueError(
+            "Please make sure that the summation of the hours is equal to 8. "
+            + f"(Given: {sum(args.task.values())} hours)"
+        )
+    logger.debug("Given task and hour pairs : %s", args.task)
```
Comment (ThorpeJosh) [inline tp_timesheet/__main__.py:90]: I'd log this before calculating the sum and potentially raising the ValueError. Will make it easier to debug.

File: tp_timesheet/__main__.py [hunk 6]
```diff
@@ -124,19 +128,16 @@ def run():
...
-            return
-
-        string_list = [f"{date}: {hours} hours" for (date, hours) in dates]
+        logger.info("Date(s) (yyyy-mm-dd) to be submitted for %s:", config.EMAIL)
         logger.info(
-            "Date(s) (yyyy-mm-dd) to be submitted for %s: %s", config.EMAIL, string_list
+            "\tWorking days : %s", [f"{date}: 8 hours" for date in working_dates]
+        )
+        logger.info(
+            "\tHolidays : %s", [f"{date}: 0 hours(holiday)" for date in holidays]
```
Comment (ThorpeJosh) [inline tp_timesheet/__main__.py:140]: These three logs are going to be slit across multiple lines. Does this look messy?

The first log implies it will have dates but only the 2nd and 3rd supply them. I'd consider rewording the log messages or combine into a single log message.

File: tp_timesheet/clockify_timesheet.py [hunk 7]
```diff
@@ -60,12 +61,14 @@ def _post_time_entry(self, date, working_hours, dry_run):
             "%Y-%m-%dT%H:%M:%SZ"
         )
 
+        project_id = self.get_project_id(task)
```
Comment (ThorpeJosh) [inline tp_timesheet/clockify_timesheet.py:64]: This adds a lot more api requests. Instead of getting the project_id once when instance is created, we are getting it every time we submit an entry. Is this avoidable? I think the build failures might be due to hitting the api requests rate limit.

File: tp_timesheet/submit_form.py [hunk 8]
```diff
@@ -61,24 +61,24 @@ def submit_timesheet(url, email, date, verbose=False, dry_run=False, working_hou
             By.XPATH,
             "(//input)[3]",
         )
-        live_hours.send_keys(working_hours)
+        live_hours.send_keys(tasks.get("live", 0))
```
Comment (ThorpeJosh) [inline tp_timesheet/submit_form.py:64]: What does the second argument do here (the `, 0`)? I thought `tasks.get("live")` or `tasks["live"]` is sufficient.

File: tp_timesheet/tests/test_clockify.py [hunk 9]
```diff
@@ -78,16 +76,19 @@ def assert_number_of_entries(test_date, number_of_entries):
...
-    clockify.submit_clockify(test_date, 1)
+    # Add a clockify entry with 4 hours
+    clockify.submit_clockify(test_date, {"live": 4})
 
     # Check only one entry exists
     assert_number_of_entries(test_date, 1)
 
+    # Remove all entries
+    clockify.delete_time_entry(test_date)
+
```
Comment (ThorpeJosh) [inline tp_timesheet/tests/test_clockify.py:87]: This test is to ensure clockify clears all previous entries when a new submission is submitted. These lines clear the entries making manually making the test pointless.

I know the mechanism for how tasks are submitted has changed so maybe this test needs to be changed. In this form it is testing nothing.

**Review by ThorpeJosh — CHANGES_REQUESTED** [review]
This is fairly large. A lot going on here. Nice work on the new interface!

**Comment by ThorpeJosh** [comment]
> This is fairly large. A lot going on here. Nice work on the new interface!

Might be a few iterations of reviews due to the complexity making it a bit hard to follow.

File: tp_timesheet/__main__.py [hunk 10]
```diff
@@ -76,13 +70,25 @@ def parse_args():
...
         help="The type of task for the clockify submission. Specify task name "
-        + "if it is anything other than 'live'. Put the task name (e.g., training, OOO, holiday)",
+        + "if it is anything other than 'live'. Put the task name(idle, training, issue, OOO, holiday).\n"
+        + 'Passing multiple task-hour pair is also acceptable. (E.g. "--task live 4 --task OOO 4" for afternoon OOO.)',
     )
-    return parser.parse_args()
+    args = parser.parse_args()
+
+    # postprocessing args
+    args.task = [["live", "8"]] if not args.task else args.task  # default value
```
Comment (siewyeng) [inline tp_timesheet/__main__.py:83]: I think this args.task should be a dictionary instead. Otherwise, it'll cause problems in the `tasks.get` of submit_form.py

**Review by siewyeng — CHANGES_REQUESTED** [review]

File: tp_timesheet/__main__.py (hunk 10 above)
Comment (siewyeng) [reply tp_timesheet/__main__.py:83]: The error was found while trying to submit hours for a holiday without indicating the task or number of hours. e.g., `tp-timesheet -s 26/12/2022 -v`. Perhaps we could find a way to add this into the tests

File: tp_timesheet/submit_form.py (hunk 8 above)
Comment (changukshin) [reply tp_timesheet/submit_form.py:64]: If user give the cli `--task training 8`, the dictionary `tasks` would be `{"training":8}`.
And when you query 'live' on it, results differ depends on the method you use.

```
tasks = {"training": 8}

# key exist
tasks.get("trainig") #8 returned

# key not exist
tasks.get("live", 0) #0 returned as "live" key does not exist on dictionary
tasks.get("live") #None returned as "live" key does not exist on dictionary
tasks["live"] #KeyError raised
```

File: tp_timesheet/__main__.py (hunk 10 above)
Comment (changukshin) [reply tp_timesheet/__main__.py:83]: This line does the job. [link](https://github.com/ThorpeJosh/tp-timesheet/pull/97/files/b3ddf3b01398e48ac2a12479aa40fc90b2f50152#diff-f313fbf2e4c858dbd3b88ac2181f480a65443050ac2b02ab120d057527ecf5dbR84)

--

I can't reproduce the error. Could you pass me the error message?

File: tp_timesheet/__main__.py (hunk 5 above)
Comment (changukshin) [reply tp_timesheet/__main__.py]: Resolved with [f65f4c0](https://github.com/ThorpeJosh/tp-timesheet/pull/97/commits/f65f4c019a6b2ff302bf47e5207028be75300f61).

File: tp_timesheet/__main__.py (hunk 6 above)
Comment (changukshin) [reply tp_timesheet/__main__.py]: Resolved with [f65f4c0](https://github.com/ThorpeJosh/tp-timesheet/pull/97/commits/f65f4c019a6b2ff302bf47e5207028be75300f61).

File: tp_timesheet/clockify_timesheet.py (hunk 7 above)
Comment (changukshin) [reply tp_timesheet/clockify_timesheet.py:67]: Resolved with [f65f4c0](https://github.com/ThorpeJosh/tp-timesheet/pull/97/commits/f65f4c019a6b2ff302bf47e5207028be75300f61).

File: tp_timesheet/tests/test_clockify.py (hunk 9 above)
Comment (changukshin) [reply tp_timesheet/tests/test_clockify.py]: Resolved with [f65f4c0](https://github.com/ThorpeJosh/tp-timesheet/pull/97/commits/f65f4c019a6b2ff302bf47e5207028be75300f61).

File: tp_timesheet/__main__.py (hunk 10 above)
Comment (siewyeng) [reply tp_timesheet/__main__.py:84]: This is the error message. Screenshot 2022-12-28 at 1 49 16 PM

**Comment by changukshin** [comment]
Check failed for Clockify-side error. Let's wait for it to be fixed.

File: tp_timesheet/__main__.py (hunk 10 above)
Comment (changukshin) [reply tp_timesheet/__main__.py:84]: Would you let me know which commit you are on? @siewyeng

Comment (siewyeng) [reply tp_timesheet/__main__.py:84]: ohhh a super old one.  b3ddf3b

Comment (changukshin) [reply tp_timesheet/__main__.py:84]: Fixed on [08dddc2](https://github.com/ThorpeJosh/tp-timesheet/pull/97/commits/08dddc2ab865ac5613e20cd796975b366b71b28d). 
Thanks! @siewyeng @chuntatlam

File: .pylintrc (hunk 2 above)
Comment (changukshin) [reply .pylintrc]: Resolved with [2294baa](https://github.com/ThorpeJosh/tp-timesheet/pull/97/commits/2294baa9792e9cf62b0724cb8a1f634f2e2d8eac).

**Comment by ThorpeJosh** [comment]
@changukshin I have rebased and resolved all the merge conflicts around the CLI parser and the project ID tests

**Comment by changukshin** [comment]
Seems all settled. Review required. @ThorpeJosh @siewyeng

**Review by ThorpeJosh — CHANGES_REQUESTED** [review]

File: tp_timesheet/clockify_timesheet.py [hunk 11]
```diff
@@ -163,39 +169,56 @@ def _get_workspace_user_id(self):
         start_time = datetime.datetime.strptime(start_time_str, "%H:%M").time()
         return workspace_id, user_id, timezone, start_time
 
-    def _get_project_id(self, task_short):
+    def get_project_id(self, task_short):
         """Send request to get project id"""
+        _, project = self.task_project_dict[task_short]
+
+        if project in self.project_id_cache:
```
[the remark this answers is not in the document]
Comment (changukshin) [reply tp_timesheet/clockify_timesheet.py:176]: Resolved with 05d10d

File: tp_timesheet/clockify_timesheet.py [hunk 12]
```diff
@@ -163,39 +169,56 @@ def _get_workspace_user_id(self):
...
             headers={"X-Api-Key": self.api_key},
             timeout=2,
         )
         get_request.raise_for_status()
         request_list = json.loads(get_request.text)
-        _, project = self.task_project_dict[task_short]
+
         for dic in request_list:
             if dic["name"] == project:
+                self.project_id_cache[project] = dic["id"]
```
[the remark this answers is not in the document]
Comment (changukshin) [reply tp_timesheet/clockify_timesheet.py:193]: Resolved with 05d10d

File: tp_timesheet/clockify_timesheet.py [hunk 13]
```diff
@@ -30,28 +33,29 @@ def __init__(self, api_key, task, locale):
...
         if not dry_run:
             # delete time entry if exist
             self.delete_time_entry(date)
 
         # post entry
-        self._post_time_entry(date, working_hours, dry_run)
+        start_time = self.start_time
+        for (task, hour) in task_and_hours.items():
+            self._post_time_entry(date, task, start_time, hour, dry_run)
+            start_time = datetime.time(start_time.hour + hour, 0)
```
[the remark this answers is not in the document]
Comment (changukshin) [reply tp_timesheet/clockify_timesheet.py]: Resolved with [3809add](https://github.com/ThorpeJosh/tp-timesheet/pull/97/commits/3809addcd3ad7b50820d920378a9ffb838517faa)

**Review by ThorpeJosh — APPROVED** [review]

File: tp_timesheet/__main__.py (hunk 10 above)
Comment (changukshin) [reply tp_timesheet/__main__.py:80]: @siewyeng Could you verify if it is resolved? I think it is okay now!
